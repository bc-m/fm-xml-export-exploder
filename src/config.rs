#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum CatalogType {
    Accounts,
    BaseDirectory,
    BaseTable,
    CalcsForCustomFunctions,
    CustomFunctions,
    CustomMenu,
    CustomMenuSet,
    ExternalDataSource,
    ExtendedPrivileges,
    FieldsForTables,
    FileAccess,
    Layout,
    Library,
    OptionsForValueLists,
    PrivilegeSets,
    Relationship,
    Script,
    StepsForScripts,
    TableOccurrence,
    Theme,
    ValueList,
}

impl CatalogType {
    pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
        match bytes {
            b"AccountsCatalog" => Some(Self::Accounts),
            b"BaseDirectoryCatalog" => Some(Self::BaseDirectory),
            b"BaseTableCatalog" => Some(Self::BaseTable),
            b"CalcsForCustomFunctions" => Some(Self::CalcsForCustomFunctions),
            b"CustomFunctionsCatalog" => Some(Self::CustomFunctions),
            b"CustomMenuCatalog" => Some(Self::CustomMenu),
            b"CustomMenuSetCatalog" => Some(Self::CustomMenuSet),
            b"ExternalDataSourceCatalog" => Some(Self::ExternalDataSource),
            b"ExtendedPrivilegesCatalog" => Some(Self::ExtendedPrivileges),
            b"FieldsForTables" => Some(Self::FieldsForTables),
            b"FileAccessCatalog" => Some(Self::FileAccess),
            b"LayoutCatalog" => Some(Self::Layout),
            b"LibraryCatalog" => Some(Self::Library),
            b"OptionsForValueLists" => Some(Self::OptionsForValueLists),
            b"PrivilegeSetsCatalog" => Some(Self::PrivilegeSets),
            b"RelationshipCatalog" => Some(Self::Relationship),
            b"ScriptCatalog" => Some(Self::Script),
            b"StepsForScripts" => Some(Self::StepsForScripts),
            b"TableOccurrenceCatalog" => Some(Self::TableOccurrence),
            b"ThemeCatalog" => Some(Self::Theme),
            b"ValueListCatalog" => Some(Self::ValueList),
            _ => None,
        }
    }

    pub fn get_config(&self) -> CatalogConfig {
        match self {
            Self::Accounts => CatalogConfig {
                catalog_item_name: b"Account".to_vec(),
                out_folder_name: "accounts".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "Account/AccountReference/@id".to_string(),
            },
            Self::BaseDirectory => CatalogConfig {
                catalog_item_name: b"BaseDirectory".to_vec(),
                out_folder_name: "base_directory".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::BaseTable => CatalogConfig {
                catalog_item_name: b"BaseTable".to_vec(),
                out_folder_name: "table_stubs".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::CalcsForCustomFunctions => CatalogConfig {
                catalog_item_name: b"CustomFunctionCalc".to_vec(),
                out_folder_name: "custom_function_calcs".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "CustomFunctionCalc/CustomFunctionReference/@id".to_string(),
            },
            Self::CustomFunctions => CatalogConfig {
                catalog_item_name: b"CustomFunction".to_vec(),
                out_folder_name: "custom_function_stubs".to_string(),
                wrapped_in_object_list: true,
                uses_folders: true,
                id_path: "".to_string(),
            },
            Self::CustomMenu => CatalogConfig {
                catalog_item_name: b"CustomMenu".to_vec(),
                out_folder_name: "custom_menus".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::CustomMenuSet => CatalogConfig {
                catalog_item_name: b"CustomMenuSet".to_vec(),
                out_folder_name: "custom_menu_sets".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::ExternalDataSource => CatalogConfig {
                catalog_item_name: b"ExternalDataSource".to_vec(),
                out_folder_name: "external_data_sources".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::ExtendedPrivileges => CatalogConfig {
                catalog_item_name: b"ExtendedPrivilege".to_vec(),
                out_folder_name: "extended_privileges".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::FieldsForTables => CatalogConfig {
                catalog_item_name: b"FieldCatalog".to_vec(),
                out_folder_name: "table_fields".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "FieldCatalog/BaseTableReference/@id".to_string(),
            },
            Self::FileAccess => CatalogConfig {
                catalog_item_name: b"Authorization".to_vec(),
                out_folder_name: "file_access".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Layout => CatalogConfig {
                catalog_item_name: b"Layout".to_vec(),
                out_folder_name: "layouts".to_string(),
                wrapped_in_object_list: false,
                uses_folders: true,
                id_path: "Layout/LayoutReference/@id".to_string(),
            },
            Self::Library => CatalogConfig {
                catalog_item_name: b"BinaryData".to_vec(),
                out_folder_name: "libraries".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "BinaryData/LibraryReference/@id".to_string(),
            },
            Self::OptionsForValueLists => CatalogConfig {
                catalog_item_name: b"ValueList".to_vec(),
                out_folder_name: "value_list_options".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "ValueList/ValueListReference/@id".to_string(),
            },
            Self::PrivilegeSets => CatalogConfig {
                catalog_item_name: b"PrivilegeSet".to_vec(),
                out_folder_name: "privilege_sets".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Relationship => CatalogConfig {
                catalog_item_name: b"Relationship".to_vec(),
                out_folder_name: "relationships".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Script => CatalogConfig {
                catalog_item_name: b"Script".to_vec(),
                out_folder_name: "script_stubs".to_string(),
                wrapped_in_object_list: false,
                uses_folders: true,
                id_path: "".to_string(),
            },
            Self::StepsForScripts => CatalogConfig {
                catalog_item_name: b"Script".to_vec(),
                out_folder_name: "script_steps".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "Script/ScriptReference/@id".to_string(),
            },
            Self::TableOccurrence => CatalogConfig {
                catalog_item_name: b"TableOccurrence".to_vec(),
                out_folder_name: "table_occurrences".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Theme => CatalogConfig {
                catalog_item_name: b"Theme".to_vec(),
                out_folder_name: "themes".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::ValueList => CatalogConfig {
                catalog_item_name: b"ValueList".to_vec(),
                out_folder_name: "value_list_stubs".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
        }
    }
}

#[derive(Debug)]
pub struct CatalogConfig {
    pub catalog_item_name: Vec<u8>,
    pub out_folder_name: String,
    pub wrapped_in_object_list: bool,
    pub uses_folders: bool,
    pub id_path: String,
}

pub struct Flags {
    pub parse_all_lines: bool,
    pub lossless: bool,
    pub verbose: Option<Option<String>>,
}
