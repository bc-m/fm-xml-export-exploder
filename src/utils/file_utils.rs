use anyhow::{anyhow, Error, Result};
use std::path::PathBuf;

/// Validate that a path exists and is a directory
pub fn valid_dir_or_throw(dir_path: &PathBuf) -> Result<(), Error> {
    let metadata = std::fs::metadata(dir_path)
        .map_err(|_| anyhow!("Path '{}' not exists", dir_path.display()))?;

    match metadata.is_dir() {
        true => Ok(()),
        false => Err(anyhow!("Path '{}' is not a directory", dir_path.display())),
    }
}

/// Join scope ID and name into a formatted string
pub fn join_scope_id_and_name(scope_id: &str, scope_name: &str) -> String {
    format!("{} - ID {}", scope_name, scope_id)
}

/// Escape filename by replacing special characters with underscores
pub fn escape_filename(filename: &str) -> String {
    let escaped_filename = filename
        .chars()
        .map(|c| match c {
            '/' | '\\' | ':' | '*' | '?' | '"' | '<' | '>' | '|' | '{' => '_',
            _ => c,
        })
        .collect::<String>();

    escaped_filename.trim().to_string()
}

/// Check if a line should be skipped based on content
pub fn should_skip_line(line: &str) -> bool {
    if line.contains("<TagList></TagList>") {
        return true;
    };
    if line.contains("<OwnerID></OwnerID>") {
        return true;
    };
    if line.contains("<Options>0</Options>") {
        return true;
    }; // Default
    if line.contains("<Options>1048576</Options>") {
        return true;
    }; // Breakpoint
    if line.contains("<UUID>") && line.contains("</UUID>") {
        return true;
    };

    false
}
