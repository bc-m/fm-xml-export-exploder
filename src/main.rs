use std::{fs, path::PathBuf, time::Instant};

use anyhow::Result;
use clap::Parser;
use rayon::prelude::*;

use crate::config::Flags;
use crate::utils::file_utils::valid_dir_or_throw;
use crate::utils::xml_utils::XmlEventType;
use crate::xml_processor::explode_xml;

mod catalog;
mod config;
mod script_sanitizer;
mod script_steps;
mod supporting;
mod tests;
mod utils;
mod xml_processor;

/// Parse all as XML exported FileMaker solutions from source directory and explode them to target directory.
#[derive(Parser)]
#[command(version, about, long_about = None)]
struct Args {
    /// The source directory to read input
    source: PathBuf,

    /// The target directory to write output
    target: PathBuf,

    /// Parse all lines (or skip less important ones to reduce noise)
    #[arg(short, long)]
    all_lines: bool,

    /// Retain all information from the main xml (or skip less important catalogs and attributes)
    #[arg(short, long)]
    lossless: bool,

    /// Enable debug/verbose output. Optionally filter by catalog name.
    /// Example: --verbose "CustomFunctionsCatalog"
    #[arg(short, long, value_name = "FILTER", num_args(0..=1))]
    verbose: Option<Option<String>>,
}

#[derive(Debug, Default)]
pub struct Skeleton {
    pub content: String,
    pub previous_line: String,
    pub previous_event_type: XmlEventType,
}

fn main() -> Result<()> {
    let start = Instant::now();

    let args = Args::parse();
    let in_dir = args.source;
    let out_dir = args.target;
    let flags = Flags {
        parse_all_lines: args.all_lines,
        lossless: args.lossless,
        verbose: args.verbose,
    };

    match &flags.verbose {
        Some(Some(filter)) => println!("🔍 Verbose mode (filtered on {})", filter),
        Some(None) => println!("🔊 Verbose mode (all)"),
        None => (), // silent
    };

    valid_dir_or_throw(&in_dir)?;

    // Read directory contents
    let paths = fs::read_dir(in_dir)?
        .filter_map(|entry| entry.ok().map(|e| e.path())) // Filter out directories and unwrap results
        .filter(|path| path.is_file() && path.extension().unwrap_or_default() == "xml") // Filter XML files
        .collect::<Vec<_>>(); // Collect paths into a vector

    println!("Start processing {} files...", paths.len());

    // Process XML files in parallel
    paths.par_iter().for_each(|path| {
        match explode_xml(path, &out_dir, &flags) {
            Ok(_) => {}
            Err(err) => {
                let file_name = path.file_name().unwrap().to_str().unwrap();
                eprintln!("Failed to process file '{}': {}", file_name, err)
            }
        };
    });

    let duration = start.elapsed();
    if duration.as_secs() > 9 {
        println!("Completed in {:?} seconds.", duration.as_secs());
    } else {
        println!("Completed in {:?} ms.", duration.as_millis());
    }

    Ok(())
}
