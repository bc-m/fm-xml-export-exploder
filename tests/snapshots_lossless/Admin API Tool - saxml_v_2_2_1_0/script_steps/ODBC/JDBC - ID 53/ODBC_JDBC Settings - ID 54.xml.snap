---
source: src/main.rs
expression: output_content
---
<Script>
	<ScriptReference id="54" name="ODBC/JDBC Settings" UUID="A61D89F9-3C36-4004-BE41-7BE82EDE7014"></ScriptReference>
	<ObjectList membercount="36">
		<Step index="0" id="89" name="# (Kommentar)" enable="True">
			<UUID>14AF30C3-BEC7-4163-B5E4-11F2577F379B</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value=" PURPOSE: Get ODBC Settings"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="1" id="89" name="# (Kommentar)" enable="True">
			<UUID>C24B7011-0ACC-4B3A-978C-96D32F7F15FA</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value=" HISTORY:&#13;   - Created:   2019-04-23 by Mike Duncan (mduncan@soliantconsulting.com)&#13;   - Modified:  2019-04-23 by Mike Duncan, update to v2 admin api"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="2" id="89" name="# (Kommentar)" enable="True">
			<UUID>7E872D7F-2694-4D32-B02D-E846D3DCF8AF</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="3" id="85" name="AnwenderAbbruchZulassen setzen" enable="True">
			<UUID>BE1F60A4-3613-4987-96D8-EAC84E2C2E83</UUID>
			<OwnerID></OwnerID>
			<Options>196608</Options>
			<ParameterValues membercount="1">
				<Parameter type="Boolean">
					<Boolean id="131072" value="True"></Boolean>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="4" id="86" name="Fehleraufzeichnung setzen" enable="True">
			<UUID>7687D3B6-F669-4365-B212-9D37C67952D9</UUID>
			<OwnerID></OwnerID>
			<Options>196608</Options>
			<ParameterValues membercount="1">
				<Parameter type="Boolean">
					<Boolean id="131072" value="True"></Boolean>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="5" id="141" name="Variable setzen" enable="True">
			<UUID>4E63CBBF-9F2D-4073-BE31-BD3B0984E207</UUID>
			<OwnerID></OwnerID>
			<Options>16388</Options>
			<ParameterValues membercount="1">
				<Parameter type="Variable">
					<value>
						<Calculation datatype="1" position="1">
							<Calculation>
								<Text><![CDATA[0]]></Text>
								<ChunkList hash="9BF55A5D51230A8EA20D4BAA0F468EEC">
									<Chunk type="NoRef">0</Chunk>
								</ChunkList>
							</Calculation>
						</Calculation>
					</value>
					<Name value="$error"></Name>
					<repetition></repetition>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="6" id="89" name="# (Kommentar)" enable="True">
			<UUID>454714D1-D438-4794-AF7D-8F32CFC6A965</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="7" id="89" name="# (Kommentar)" enable="True">
			<UUID>EC8C9C00-B14B-4D52-9ED7-8BDC32B278EE</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="-------------------------------------------------------------"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="8" id="89" name="# (Kommentar)" enable="True">
			<UUID>8E2B7242-28B1-4747-BD38-F59603202B50</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value=" SET-UP"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="9" id="89" name="# (Kommentar)" enable="True">
			<UUID>2D00C912-9068-4668-8797-5C381B5A3354</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="-------------------------------------------------------------"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="10" id="1" name="Script ausführen" enable="True">
			<UUID>8B8B81FD-04E9-46F1-A774-65535B6FAB5E</UUID>
			<OwnerID></OwnerID>
			<Options>64</Options>
			<ParameterValues membercount="2">
				<Parameter type="List">
					<List name="Aus Liste" value="1">
						<ScriptReference id="35" name="Set Config" UUID="6E806D84-9200-44AE-B032-89A6016326FF"></ScriptReference>
					</List>
				</Parameter>
				<Parameter type="Parameter">
					<Parameter></Parameter>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="11" id="141" name="Variable setzen" enable="True">
			<UUID>39C7BBC5-E632-4AC6-8202-46A7B3C96A2E</UUID>
			<OwnerID></OwnerID>
			<Options>16388</Options>
			<ParameterValues membercount="1">
				<Parameter type="Variable">
					<value>
						<Calculation datatype="1" position="1">
							<Calculation>
								<Text><![CDATA["/xdbc/config"]]></Text>
								<ChunkList hash="B1EFEC4F515391208565F47C368B74E5">
									<Chunk type="NoRef">&quot;/xdbc/config&quot;</Chunk>
								</ChunkList>
							</Calculation>
						</Calculation>
					</value>
					<Name value="$endpoint"></Name>
					<repetition></repetition>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="12" id="141" name="Variable setzen" enable="True">
			<UUID>A07D14E8-725A-4546-9830-D232531BF69A</UUID>
			<OwnerID></OwnerID>
			<Options>16388</Options>
			<ParameterValues membercount="1">
				<Parameter type="Variable">
					<value>
						<Calculation datatype="1" position="1">
							<Calculation>
								<Text><![CDATA[" --request GET" &" --header \"Authorization: Bearer " & $$my.token & "\"" &" --header \"Content-Type: application/json\"" &" --dump-header $$my.headers"]]></Text>
								<ChunkList hash="73E91CCB3C96E1EE23F539872E85DAF1">
									<Chunk type="NoRef">&quot; --request GET&quot; &amp;&#13;&quot; --header \&quot;Authorization: Bearer &quot; &amp; </Chunk>
									<Chunk type="VariableReference">$$my.token</Chunk>
									<Chunk type="NoRef"> &amp; &quot;\&quot;&quot; &amp;&#13;&quot; --header \&quot;Content-Type: application/json\&quot;&quot; &amp;&#13;&quot; --dump-header $$my.headers&quot;</Chunk>
								</ChunkList>
							</Calculation>
						</Calculation>
					</value>
					<Name value="$curloptions"></Name>
					<repetition></repetition>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="13" id="89" name="# (Kommentar)" enable="True">
			<UUID>544CDC18-0530-446C-A052-6E6E9B59D908</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="14" id="89" name="# (Kommentar)" enable="True">
			<UUID>2EDFA3C8-1151-49E8-A60D-41C8FD7EF32E</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="-------------------------------------------------------------"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="15" id="89" name="# (Kommentar)" enable="True">
			<UUID>AC173346-367B-43E6-AD25-53E7F18DCEF6</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value=" PROCESS"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="16" id="89" name="# (Kommentar)" enable="True">
			<UUID>C65D07F9-61AF-40CB-9C5E-A86B4E8568EB</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="-------------------------------------------------------------"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="17" id="160" name="Aus URL einfügen" enable="True">
			<UUID>4430108B-5E54-4FCD-9A06-F66C4B6A6E1F</UUID>
			<OwnerID></OwnerID>
			<Options>20613</Options>
			<ParameterValues membercount="6">
				<Parameter type="Boolean">
					<Boolean type="SSL-Zertifikate verifizieren" id="268435456" value="False"></Boolean>
				</Parameter>
				<Parameter type="Boolean">
					<Boolean type="Auswahl" id="4096" value="True"></Boolean>
				</Parameter>
				<Parameter type="Boolean">
					<Boolean type="Mit Dialog" id="128" value="False"></Boolean>
				</Parameter>
				<Parameter type="Target">
					<Variable value="$$my.result">
						<repetition value="1"></repetition>
					</Variable>
				</Parameter>
				<Parameter type="URL">
					<URL autoEncode="True">
						<Calculation datatype="1" position="0">
							<Calculation>
								<Text><![CDATA[$$server & $endpoint]]></Text>
								<ChunkList hash="9EA92AEF077FA44AE68308694556F8CB">
									<Chunk type="VariableReference">$$server</Chunk>
									<Chunk type="NoRef"> &amp; </Chunk>
									<Chunk type="VariableReference">$endpoint</Chunk>
								</ChunkList>
							</Calculation>
						</Calculation>
					</URL>
				</Parameter>
				<Parameter type="Calculation">
					<Calculation datatype="1" position="1">
						<Calculation>
							<Text><![CDATA[$curloptions]]></Text>
							<ChunkList hash="205B4DA1BAD160DEDA31FE70167BBA78">
								<Chunk type="VariableReference">$curloptions</Chunk>
							</ChunkList>
						</Calculation>
					</Calculation>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="18" id="1" name="Script ausführen" enable="True">
			<UUID>C4BC245D-9769-4769-87A3-3B718B7FD9A5</UUID>
			<OwnerID></OwnerID>
			<Options>64</Options>
			<ParameterValues membercount="2">
				<Parameter type="List">
					<List name="Aus Liste" value="1">
						<ScriptReference id="31" name="Handle response" UUID="44FA18ED-6DEA-4F32-BDDB-09F438F6D005"></ScriptReference>
					</List>
				</Parameter>
				<Parameter type="Parameter">
					<Parameter></Parameter>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="19" id="89" name="# (Kommentar)" enable="True">
			<UUID>45E5737C-7F8B-4670-BB99-996534ADCFCB</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="20" id="68" name="Wenn" enable="True">
			<UUID>D39D6F13-75A4-421D-9EC3-D7BD40A8E44D</UUID>
			<OwnerID></OwnerID>
			<Options>16384</Options>
			<ParameterValues membercount="1">
				<Parameter type="Calculation">
					<Calculation datatype="7" position="0">
						<Calculation>
							<Text><![CDATA[Hole ( ScriptErgebnis ) = "1"]]></Text>
							<ChunkList hash="8C6CBCB21CCFFBA5E461DF278EC4FA0D">
								<Chunk type="FunctionRef">Hole</Chunk>
								<Chunk type="NoRef"> ( </Chunk>
								<Chunk type="FunctionRef">ScriptErgebnis</Chunk>
								<Chunk type="NoRef"> ) = &quot;1&quot;</Chunk>
							</ChunkList>
						</Calculation>
					</Calculation>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="21" id="89" name="# (Kommentar)" enable="True">
			<UUID>8BB8EBC4-D9E5-410F-B7CD-6464F5D8A289</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="parse json response"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="22" id="76" name="Feldwert setzen" enable="True">
			<UUID>BDC4853C-9E96-432C-99C3-DAB78B70DF9A</UUID>
			<OwnerID></OwnerID>
			<Options>16385</Options>
			<ParameterValues membercount="2">
				<Parameter type="FieldReference">
					<FieldReference id="40" name="odbc_enabled" UUID="42186240-A433-45B4-85CB-303AD3BA5A72">
						<repetition value="1"></repetition>
						<TableOccurrenceReference id="1065089" name="FMS__fmserver" UUID="597569C5-24CD-486D-84D6-D79B9D3BDD5B"></TableOccurrenceReference>
					</FieldReference>
				</Parameter>
				<Parameter type="Calculation">
					<Calculation datatype="1" position="0">
						<Calculation>
							<Text><![CDATA[JSONGetElement ( $$my.result ; "response.enabled" )]]></Text>
							<ChunkList hash="29A2524F996FD70658961FEA97F52F1E">
								<Chunk type="FunctionRef">JSONGetElement</Chunk>
								<Chunk type="NoRef"> ( </Chunk>
								<Chunk type="VariableReference">$$my.result</Chunk>
								<Chunk type="NoRef"> ; &quot;response.enabled&quot; )</Chunk>
							</ChunkList>
						</Calculation>
					</Calculation>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="23" id="89" name="# (Kommentar)" enable="True">
			<UUID>CA0EC9C2-CEAC-4294-B9BE-E2EEC7F504E5</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="24" id="70" name="Ende (wenn)" enable="True">
			<UUID>2C4E1358-02D6-452D-BF2E-334C1ADF61F9</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
		</Step>
		<Step index="25" id="89" name="# (Kommentar)" enable="True">
			<UUID>28E46E67-9352-48DD-850F-E8EC8C913DA0</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="26" id="89" name="# (Kommentar)" enable="True">
			<UUID>DC1278F8-914C-4731-B508-2C4990A2127C</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="-------------------------------------------------------------"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="27" id="89" name="# (Kommentar)" enable="True">
			<UUID>C46F4460-4B8E-4170-B976-FC5990D627C8</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value=" CLEAN UP"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="28" id="89" name="# (Kommentar)" enable="True">
			<UUID>F971FC20-1F5A-4472-AECA-58A5EBB26E4C</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="-------------------------------------------------------------"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="29" id="68" name="Wenn" enable="True">
			<UUID>CACB8678-6AD9-48F2-B061-A79A243AA279</UUID>
			<OwnerID></OwnerID>
			<Options>16384</Options>
			<ParameterValues membercount="1">
				<Parameter type="Calculation">
					<Calculation datatype="7" position="0">
						<Calculation>
							<Text><![CDATA[$error = 0]]></Text>
							<ChunkList hash="CE3C173E2CD2554CBB236CEE7C56CBB6">
								<Chunk type="VariableReference">$error</Chunk>
								<Chunk type="NoRef"> = 0</Chunk>
							</ChunkList>
						</Calculation>
					</Calculation>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="30" id="89" name="# (Kommentar)" enable="True">
			<UUID>85A1E358-F14E-46A9-9A68-202B48B5A372</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value=" set to exit true if you do not need to pass an error"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="31" id="141" name="Variable setzen" enable="True">
			<UUID>1FCAC33E-511E-4A8D-98B0-555A5FE3FB3B</UUID>
			<OwnerID></OwnerID>
			<Options>16388</Options>
			<ParameterValues membercount="1">
				<Parameter type="Variable">
					<value>
						<Calculation datatype="1" position="1">
							<Calculation>
								<Text><![CDATA[1 // exit true]]></Text>
								<ChunkList hash="DC134DF247804ECF23193E2024A59A1B">
									<Chunk type="NoRef">1 // exit true</Chunk>
								</ChunkList>
							</Calculation>
						</Calculation>
					</value>
					<Name value="$error"></Name>
					<repetition></repetition>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="32" id="70" name="Ende (wenn)" enable="True">
			<UUID>804AA0B5-10F4-4D85-944E-6B025DF0A508</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
		</Step>
		<Step index="33" id="89" name="# (Kommentar)" enable="True">
			<UUID>9208A74A-EC2C-4532-9CA7-4C1A7A3FD040</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="34" id="103" name="Aktuelles Script verlassen" enable="True">
			<UUID>59298D56-429F-497E-9ACB-8546354633F5</UUID>
			<OwnerID></OwnerID>
			<Options>16384</Options>
			<ParameterValues membercount="1">
				<Parameter type="Calculation">
					<Calculation datatype="1" position="0">
						<Calculation>
							<Text><![CDATA[$error]]></Text>
							<ChunkList hash="926F59D0E89012933FB1C2137ABF78F1">
								<Chunk type="VariableReference">$error</Chunk>
							</ChunkList>
						</Calculation>
					</Calculation>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="35" id="89" name="# (Kommentar)" enable="True">
			<UUID>2F77C3AE-8763-4EB4-9A41-6EDC32542D3B</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
	</ObjectList>
</Script>
